### ISSUE TITLE
[MSHADE-36] Add option to include dependency reduced POM instead of original one

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#404]
**[Mark Hobson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mihobson)** opened **[MSHADE-36](https://issues.apache.org/jira/browse/MSHADE-36?redirect=false)** and commented

The dependency reduced pom needs to replace the packaged pom at:

META-INF/maven/\<groupId>/\<artifactId>/pom.xml

This will allow tools that use this metadata to read the correct shaded dependencies (e.g. maven-runtime).

---

**Affects:** 1.1

**Issue Links:**
- [MSHADE-82](https://issues.apache.org/jira/browse/MSHADE-82) Add the ability to deploy a reduced pom in place of  the original
 (_**"duplicates"**_)
- [MSHADE-82](https://issues.apache.org/jira/browse/MSHADE-82) Add the ability to deploy a reduced pom in place of  the original

**Remote Links:**
- [GitHub Pull Request #144
](https://github.com/apache/maven-shade-plugin/pull/144)
- [GitHub Pull Request #25
](https://github.com/apache/maven-shade-plugin/pull/25)

8 votes, 14 watchers

**Comment by jira-importer** [comment]
**[Robert Burrell Donkin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=robertburrelldonkin)** commented

Would creating a ResourceTransformer be the best approach?

**Comment by jira-importer** [comment]
**[Hung Huynh](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hhuynh)** commented

I prefer this as a built-in default feature since it's expected rather an exception. However, users should be able to turn this off if they want to

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This issue has been auto closed because it has been inactive for a long period of time. If you think this issue still persists, retest your problem with the most recent version of Maven and the affected component, reopen and post your results.

**Comment by jira-importer** [comment]
**[Volkan Yazici](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vy)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), I think this is still a valid bug. Here is a test case to reproduce it:

This is how my `pom.xml` looks like:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.vlkan</groupId>
    <artifactId>maven-shade-plugin-pom-bug</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna</artifactId>
            <version>4.2.1</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-shade-plugin</artifactId>
                <version>2.4.3</version>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>shade</goal>
                        </goals>
                        <configuration>
                            <artifactSet>
                                <includes>
                                    <include>net.java.dev.jna:jna</include>
                                </includes>
                            </artifactSet>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

</project>
```

Here, since I am including `net.java.dev.jna:jna` within the final artifact, I would expect it to be removed from the `META-INF/maven/com.vlkan/maven-shade-plugin-pom-bug/pom.xml` file. But apparently, it is not:

```
$ mvn package
$ unzip -p target/maven-shade-plugin-pom-bug-1.0-SNAPSHOT.jar META-INF/maven/com.vlkan/maven-shade-plugin-pom-bug/pom.xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.vlkan</groupId>
    <artifactId>maven-shade-plugin-pom-bug</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna</artifactId>
            <version>4.2.1</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-shade-plugin</artifactId>
                <version>2.4.3</version>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>shade</goal>
                        </goals>
                        <configuration>
                            <artifactSet>
                                <includes>
                                    <include>net.java.dev.jna:jna</include>
                                </includes>
                            </artifactSet>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

</project>
```

Per see, `net.java.dev.jna:jna` still appears as a dependency, which actually is not any more. It should have been removed from the dependencies of the final artifact.

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), can we re-open this issue? (I suppose I am lacking the sufficient rights to do so.)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Reopening per user request.

**Comment by jira-importer** [comment]
**[Jonathan S Fisher](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=exabrial)** commented

Does anyone have a patch for this issue? If not, where would be a good place to start bug hunting?

**Comment by jira-importer** [comment]
**[Manu Manjunath](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=m-manu)** commented

Can you guys at least provide us a flag in shade plugin like this:

```xml
<useDependencyReducedPomInJar>true</useDependencyReducedPomInJar>
```

When set to true, the `jar` that is built should contain `pom.xml` whose content is same `dependency-reduced-pom.xml` that shade  plugin generates. That way:
1. Backword compatibility is maintained
2. Developers can keep this false if some of the dependent code is dynamically loading classes, using classloader.

**Comment by jira-importer** [comment]
**[Ben Tatham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentatham@nanometrics.ca)** commented

Would fixing this allow the dependency-reduced-pom to be used as the pom of the module in a reactor build?

**Comment by jira-importer** [comment]
**[Ben Tatham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentatham@nanometrics.ca)** commented

Would fixing this allow the dependency-reduced-pom to be used as the pom of the module in a reactor build?

**Comment by jira-importer** [comment]
**[Niels Basjes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=nielsbasjes)** commented

I put up a pull request with a proposed fix for this:

https://github.com/apache/maven-shade-plugin/pull/25

**Comment by jira-importer** [comment]
**[Niels Basjes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=nielsbasjes)** commented

[Ben Tatham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentatham@nanometrics.ca) : When building a multi module project there will still be problems regarding the dependencies in the other modules.

I recently wrote a page about that how I solved that in my own project: https://yauaa.basjes.nl/NOTES-shading-dependencies.html

Essentially what works for me:
* Library has dependencies (shaded into the jar)
* Other modules use it as a dependency, exclude the shaded dependencies, re-include the shaded dependencies as 'provided'.

**Comment by jira-importer** [comment]
**[Alexey Subach](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alexey.subach)** commented

Dear Team,
I think this is a terribly useful feature - is there any chance it can be given a bit more priority?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Waiting for the PR to be updated...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [83c123d1f9c5f6927af2aca12ee322b5168a7c63](https://gitbox.apache.org/repos/asf?p=maven-shade-plugin.git&a=commit&h=83c123d1f9c5f6927af2aca12ee322b5168a7c63).

**Comment by jira-importer** [comment]
**[John Dimeo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdimeo)** commented

Thanks for the work on this, but something very very strange just happened: one of my child modules of one of my multi-module projects seemigly used the new dep reduced POM as it's "real" POM and it then removed all transitive dependencies (without me telling the plugin to do so) for downstream Maven builds. I'm not sure how or why it happened, but just a heads up.  From my perspective, I just updated the Maven shade plugin and transitive dependencies stopped working- no configuration was changed. The workaround is to set `useDependencyReducedPomInJar` back to false (since thankfully no one downstream is using it).
